aten/src/ATen/native/Distributions.h

```diff
@@ -0,0 +1,215 @@
+  scalar_t scale = 1.0;
+
+  // Boost alpha for higher acceptance probability.
+  if (alpha < 1.0) {
+    scale *= std::pow(1 - standard_uniform.sample(), static_cast<scalar_t>(1.0) / alpha);
+    alpha += 1.0;
+  }
+
+  // This implements the acceptance-rejection method of Marsaglia and Tsang (2000)
+  // doi:10.1145/358407.358414
+  const scalar_t d = alpha - 1.0 / 3.0;
+  const scalar_t c = 1.0 / std::sqrt(9.0 * d);
+  for (;;) {
+    scalar_t x, y;
+    do {
+      x = standard_normal.sample();
+      y = 1.0 + c * x;
+    } while (y <= 0);
+    const scalar_t v = y * y * y;
+    const scalar_t u = 1 - standard_uniform.sample();
+    const scalar_t xx = x * x;
+    if (u < 1.0 - 0.0331 * xx * xx)
+      return scale * d * v;
+    if (std::log(u) < 0.5 * xx + d * (1.0 - v + std::log(v)))
+      return scale * d * v;
+  }
+}
+
+template <typename scalar_t>
+deviceforcuda static inline scalar_t polevl(const scalar_t x,  const scalar_t A[], size_t len) {
+  scalar_t result = 0;
+  for (size_t i = 0; i <= len; i++) {
+    result = result * x + A[i];
+  }
+  return result;
+}
+
+
+/*
+ * The following function comes with the following copyright notice.
+ * It has been released under the BSD license.
+ *
+ * Cephes Math Library Release 2.8:  June, 2000
+ * Copyright 1984, 1987, 1992, 2000 by Stephen L. Moshier
+ */
+template <typename scalar_t>
+deviceforcuda static inline scalar_t digamma_one(scalar_t x) {
+  using acc_scalar_t = typename std::conditional<std::is_same<scalar_t, at::Half>::value, float, scalar_t>::type;
+  constexpr acc_scalar_t PSI_10 = 2.25175258906672110764;
+  if (x == 0) {
+    return INFINITY;
+  }
+  acc_scalar_t additional_summand = 0;
+  int x_is_integer = x == std::floor(x);
+  if (x < 0) {
+    if (x_is_integer) {
+      return INFINITY;
+    }
+    // it is more standard to write this as recursion, but
+    // nvcc does not like that
+    additional_summand = - M_PI / std::tan(M_PI * x);
+    x = 1 - x;
+  }
+
+  // Push x to be >= 10
+  acc_scalar_t result = 0;
+  while (x < 10) {
+    result -= 1 / x;
+    x += 1;
+  }
+  if (x == 10) {
+    return result + PSI_10 + additional_summand;
+  }
+
+  // Compute asymptotic digamma
+  static const acc_scalar_t A[] = {
+     8.33333333333333333333E-2,
+    -2.10927960927960927961E-2,
+     7.57575757575757575758E-3,
+    -4.16666666666666666667E-3,
+     3.96825396825396825397E-3,
+    -8.33333333333333333333E-3,
+     8.33333333333333333333E-2,
+  };
+
+  acc_scalar_t y = 0;
+  if (x < 1.0e17) {
+    acc_scalar_t z = 1.0 / (x * x);
+    y = z * polevl<acc_scalar_t>(z, A, 6);
+  }
+  return static_cast<scalar_t>(result + std::log(x) - (0.5 / x) - y + additional_summand);
+}
+
+// Computes the reparameterized gradient -(d/dalpha cdf(x;alpha)) / pdf(x;alpha)
+// for random number x drawn from a standard Gamma distribution Gamma(alpha).
+template <typename scalar_t>
+deviceforcuda scalar_t standard_gamma_grad_one(scalar_t alpha, scalar_t x) {
```

Comment:
Can we compute this whole function with at least single precision?

Reply:
One way to do this is to cast to float at the call site inside `gamma_grad_cuda_kernel`, i.e. cast `self_val` and `output_val` to float before calling this. I think it's safe to say that no other functions will call this, and that way we can 'upgrade' this function with something that works in half-precision if that comes along.